### PR TITLE
[#1994] Migrate build.gradle from Groovy to Kotlin 

### LINKED ISSUE #1994: Gradle: Migrate from Groovy to Kotlin
### What feature(s) would you like to see in RepoSense

Our current build.gradle script uses the Groovy syntax (example below):
```
plugins {
    id 'application'
    id 'checkstyle'
    id 'idea'
    id 'jacoco'
    id 'java'
    id 'com.github.johnrengelman.shadow' version '7.1.2'
    id 'com.liferay.node' version '7.2.18'
    id 'com.github.psxpaul.execfork' version '0.2.0'
    id 'com.palantir.git-version' version '0.13.0'
}
```

### Is the feature request related to a problem?

Kotlin has been recommended due to cleaner syntax and enhanced editing experience in IDEs, particularly IntelliJ. Below shows the equivalent block in Kotlin syntax:
```
plugins {
    application
    checkstyle
    idea
    jacoco
    java
    id("com.github.johnrengelman.shadow") version "7.1.2"
    id("com.liferay.node") version "7.2.18"
    id("com.github.psxpaul.execfork") version "0.2.0"
    id("com.palantir.git-version") version "0.13.0"
}
```

One key advantage of Kotlin, imo, is that the syntax is much closer to that of Java, which we use for the backend, compared to Groovy. As an example, Groovy allows single and double quotes for strings whereas Kotlin requires strings to be in double-quotes, just like in Java.

Having a more standardised syntax means that someone who starts work on build.gradle.kts will find it easier to transition to the Java code.

### If possible, describe the solution

Useful links:
* https://docs.gradle.org/current/userguide/kotlin_dsl.html
* https://docs.gradle.org/current/userguide/migrating_from_groovy_to_kotlin_dsl.html

### PR DESCRIPTION
**Pull request by nseah21**
Fixes #1994 

### Migrate build.gradle from Groovy to Kotlin

```
The current build script uses Groovy syntax. 

Kotlin has been recommended, however, due to cleaner syntax 
and a better editing experience, owing to its similarity with Java.

Let's refactor the build.gradle file to use Kotlin instead, so that
someone who starts work on build.gradle.kts will find it easier 
to transition to the Java code.
```

### Other information

It is worth noting that builds using Kotlin tend to be slower than builds using Groovy. 

However, it is assumed that the added benefit of a more standardised syntax (as mentioned in #1994) outweighs this performance tradeoff.

### CODE REVIEW

**Comment by nseah21** [comment]
When running `gradlew clean build` in root directory, the build terminates with an exception after executing the `:installFrontend` task. 

Could I kindly request for some guidance on how to resolve the following issue?
```
> Task :installFrontend FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':installFrontend'.
> Process 'command 'cmd'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.6.1/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 28s
9 actionable tasks: 9 executed
```

**Comment by sikai00** [comment]
I did a quick run of your PR changes with `gradlew clean build --scan --info`:

image

vs. branch master:

image

It looks like the args `ci` is not passed as an argument for some reason. Maybe you can start from there?

**Comment by sikai00** [comment]
I ran the changes in your PR with the `scan` and `info` flag:

I got this:
image

```
Starting process 'command '/Users/sikai/RepoSense/build/node/bin/node''. Working directory: /Users/sikai/RepoSense/frontend Command: /Users/sikai/RepoSense/build/node/bin/node /Users/sikai/RepoSense/build/node/lib/node_modules/npm/bin/npm-cli.js ci --production false --cache /Users/sikai/RepoSense/frontend --loglevel info --progress true
Successfully started process 'command '/Users/sikai/RepoSense/build/node/bin/node''
npm info it worked if it ends with ok
npm info using npm@3.3.12
npm info using node@v5.5.0

Usage: npm <command>
```

versus master branch:
image

This might be a good starting point. Seems like the arguments were not passed into npm directly but rather through `/Users/sikai/RepoSense/build/node/lib/node_modules/npm/bin/npm-cli.js ci --production false --cache `.

**Comment by nseah21** [comment]
@sikai00 Thank you for pointing me in the right direction. I have re-attempted the migration and have gotten most of the tasks to run successfully.

However, the task `:run` seems to have broken, as invoking `gradlew run -Dargs="--view --repo <repo-url>"` results in the following:

```
> Task :run
02:21:42 - Config path not provided, using the config folder as default.
02:21:42 - Parsed header of CSV file, repo-config.csv, and recognized columns: ...
```

instead of:

```
> Task :run
02:24:29 - Log temp folder has been successfully created
02:24:30 - Cloning in parallel from <repo-url>...
```

when building with the original `build.gradle` file.

Has anyone experienced this issue previously when writing the `build.gradle` in Groovy?

**Comment by gok99** [comment]
Changes are looking good so far! Don't forget the newline at EOF so that the CI tests can run.

File: build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+    finalizedBy(linuxHotReloadFrontend)
+}
+
+fun deleteReposAddressDirectory() {
+    val REPOS_ADDRESS = "repos"
+    val reposDirectory = File(REPOS_ADDRESS)
+    reposDirectory.deleteRecursively()
+}
+
+defaultTasks("clean", "build", "systemTest", "frontendTest", "coverage")
```
Comment (gok99) [inline build.gradle.kts:399]:
```suggestion
defaultTasks("clean", "build", "systemTest", "frontendTest", "coverage")

```

**Comment by gok99** [comment]
Bumping the failing build: `setCommandLine("cmd")` wouldn't work on Linux and Mac. @nseah21 Have you tried using the liferay node plugin we used in the groovy build.gradle?

**Comment by yhtMinceraft1010X** [comment]
@nseah21 Have you tried running `./gradlew clean checkstyleAll test systemTest coverage` (must be this exact command, iirc `gradle clean build` won't run the `coverage` task) with `--info` or `--debug` flags on your local machine?

The additional logging info might help in determining the exact cause behind `Unable to read execution data file /Users/runner/work/RepoSense/RepoSense/build/test-results/systemtest/binary`

File: build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+        csv.required.set(true)
+        html.destination = file("${buildDir}/jacocoHtml")
+    }
+    executionData(systemtest, frontendTest)
+}
+
+tasks.register<JacocoReport>("coverage")
+
+tasks.named<JacocoReport>("coverage").configure {
+    dependsOn(systemtest, frontendTest)
```
Comment (yhtMinceraft1010X) [inline build.gradle.kts:313]: Task dependency should not be introduced here

**Comment by nseah21** [comment]
I am currently troubleshooting some of the issues in this PR. 

I have reached out to other contributors to discuss the fixes.

**Comment by sikai00** [comment]
Per the error message,

```
Gradle detected a problem with the following location: 'D:\a\RepoSense\RepoSense\build\jacoco\test.exec'. Reason: Task ':coverage' uses this output of task ':test' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.
```

it seems the fix to your issue is to declare an explicit dependency for the predefined task 'test'. This follows from coverage using 'test' to know the amount of coverage.

As for the second issue, I did some Googling online. Perhaps [this](https://github.com/gradle/gradle/issues/5898) may help?

Hope this helps you fix the issues you are facing!

**Comment by MarcusTXK** [comment]
Hi @nseah21, just bumping this PR to check its progress. Could I check if there are any major blockers for you currently?

**Comment by nseah21** [comment]
Hi @MarcusTXK, I am still trying to troubleshoot an issue with reading the execution data file. 

```
* What went wrong:
Execution failed for task ':coverage'.
> Unable to read execution data file <file_path>\RepoSense\build\test-results\systemtest\binary
```

I have attempted the fix posted by @sikai00 above but the build is still throwing the error. 

May I request for some guidance on this issue? Thank you.

**Comment by damithc** [comment]
@nseah21 I'm not sure if we want to merge this PR in the end, as I mentioned in https://github.com/reposense/RepoSense/issues/1994#issuecomment-1784041715
Even if we don't merge it in the end, it is still fine to continue until we reach a LGTM state so that at least we know what it would look like.

**Comment by nseah21** [comment]
@gok99 @yhtMinceraft1010X This pull request is ready for the next round of reviews.

**Pull request closed without merging**